tensorizer/serialization.py

```diff
@@ -3700,529 +3766,152 @@ def write_tensor(
-            ) = self._async_bulk_device_to_host_transfer(cuda_tensors)
-        else:
-            transferred = interrupt_transfer = None
-        del cuda_tensors
+            self._synchronize_pools()
+            self._sync_prologue_state()
+        except Exception as e:
+            for j in self._jobs:
+                j.cancel()
+            if cuda_executor is not None:
+                cuda_executor.shutdown(wait=False)
+            raise e
 
-        if self._encrypted:
-            shared = []
-            seen_addresses = set()
-            for t in reversed(tensors):
-                if t.tensor.device.type in ("cuda", "meta"):
-                    shared.append(False)
-                else:
-                    address = t.tensor.data_ptr()
-                    shared.append(address in seen_addresses)
-                    seen_addresses.add(address)
-            del seen_addresses
-        else:
-            shared = [False] * len(tensors)
-
-        try:
-            while tensors:
-                idx, name, tensor_type, tensor, callback = tensors.popleft()
-                is_shared = shared.pop()
-                self._idx = idx
-                if tensor.device.type == "cuda":
-                    tensor = next(transferred)
-                    temp_tensor = True
-                elif is_shared and self._encrypted:
-                    # Un-shares tensor memory in preparation for in-place
-                    # operations on the buffer that would otherwise conflict
-                    # with one another. Full support for shared-memory tensors
-                    # (e.g. if they were only written once) could make
-                    # this unnecessary, once implemented.
-                    # Another option would be to reuse the same encrypted
-                    # weights and decrypt them at the end. This would require
-                    # confirming that the tensor data regions are actually
-                    # identical, and don't just overlap.
-                    tensor = tensor.clone().detach()
-                    temp_tensor = True
-                else:
-                    temp_tensor = False
-                next_pos = self._write_tensor(
-                    idx,
-                    name,
-                    tensor_type,
-                    tensor,
-                    _synchronize=False,
-                    _start_pos=next_pos,
-                    _temporary_buffer=temp_tensor,
-                )
-                if callback is not None:
-                    callback()
-        except Exception:
-            if interrupt_transfer is not None:
-                interrupt_transfer()
-            raise
-        self._synchronize_pools()
-        self._file.seek(next_pos)
-        self._sync_prologue_state()
+        if cuda_executor is not None:
+            cuda_executor.shutdown(wait=True)
 
     def write_module(
         self,
         m: torch.nn.Module,
-        remove_tensors: bool = False,
```

Comment:
What are the consequences of removing this argument? Do we know if anyone is using this?

Reply:
I don't really know why this was useful imo, but open to pushback. In a major version update we can break the interface

Reply:
Once upon a time, this could potentially progressively deallocate memory by consuming a module, but it was sort of at the whims of the garbage collector, and wouldn't really reduce peak RAM usage anyway, so it wasn't very useful. This used to be the default behaviour, and the first time I started engaging with the tensorizer repo was to help someone debug this behaviour, because it was unintuitive and messing their stuff up. I don't see too much value to keeping it, especially since parallelized bulk writing tends to keep references to tensors around for much longer.

If anyone does need progressive deallocation of memory, this would probably be better implemented as a generalized callback hook for when each tensor has finished being processed, which would let other code deallocate tensor memory *and actually know that it has happened so that they can start taking advantage of it*... among other use cases. Since this is a blocking interface, it's currently unlikely that anyone could use it more effectively than they could write `del m` after the call returns, so the workaround for any code currently using it just to free memory at some nonspecific point shouldn't be terribly inconvenient.